tools/wptrunner/wptrunner/executors/executorchrome.py

```diff
@@ -57,12 +61,93 @@ def convert_from_crashtest_result(test, result):
 _SanitizerMixin = make_sanitizer_mixin(WebDriverCrashtestExecutor)
 
 
-class ChromeDriverRefTestExecutor(WebDriverRefTestExecutor, _SanitizerMixin):  # type: ignore
-    pass
+class ChromeDriverTestharnessProtocolPart(WebDriverTestharnessProtocolPart):
+    """Implementation of `testharness.js` tests controlled by ChromeDriver.
 
+    The main difference from the default WebDriver testharness implementation is
+    that the test window can be reused between tests for better performance.
+    """
 
-class ChromeDriverTestharnessExecutor(WebDriverTestharnessExecutor, _SanitizerMixin):  # type: ignore
-    pass
+    def setup(self):
+        super().setup()
+        # Handle (an alphanumeric string) that may be set if window reuse is
+        # enabled. This state allows the protocol to distinguish the test
+        # window from other windows a test itself may create that the "Get
+        # Window Handles" command also returns.
+        #
+        # Because test window persistence is a Chrome-only feature, it's not
+        # exposed to the base WebDriver testharness executor.
+        self.test_window = None
+        self.reuse_window = self.parent.reuse_window
+
+    def load_runner(self, url_protocol):
+        prev_runner_handle = self.runner_handle
+        super().load_runner(url_protocol)
+        if self.runner_handle != prev_runner_handle:
```

Comment:
In what situation this will be true? Also at line 107 to close a window, it set the window handle to the one to be closed. Does this undo what is done in "super().load_runner"?

Reply:
How about we set test_window to None, then close_old_windows will automatically close old test window?

Reply:
(Was going to reply after checking the stability of [this change](https://ci.chromium.org/ui/p/chromium/builders/try/linux-wpt-fyi-rel/1895/overview) in Chromium CI.)

> In what situation this will be true?

Good question. Looking more closely at `WebDriverTestharnessExecutor.load_runner`, `runner_handle` should actually be the same for the lifetime of the executor:

https://github.com/web-platform-tests/wpt/blob/3408d1c3007655d458c5b12650e6aec8f7f066a3/tools/wptrunner/wptrunner/executors/executorwebdriver.py#L106-L116

The exchange between `self.runner_handle` and `self.webdriver.window_handle` is an initialization mechanism to get the session's initial window handle (i.e., first tab) into `runner_handle` when it's initially `None`. After that, the `runner_handle` shouldn't change, even if `testharness_runner.html` is loaded with a different protocol. Therefore, this override is functionally a no-op and can be removed.

(Also, as you noted, the active `window_handle` was different when leaving the overridden `load_runner`, but that postcondition didn't matter to `do_testharness`.)

Reply:
Another unrelated question is how often do we have a protocol change, thus need to call load_runner? And how much time is used due to this? We can potentially save some time if tests are also ordered by protocol.

`load_runner` does two things: 1) load testharness_runner.html, 2)set the title of the window. Why we need to redo those things when protocol change?

Reply:
I think these are better questions for someone who knows the testharness internals better, but some approximate answers:

> is how often do we have a protocol change, thus need to call load_runner? And how much time is used due to this?

Downstream, I tried

```
./run_wpt_tests.py -t Default content-security-policy --product=chrome --reuse-window -v --processes=1
```

and saw 222 `load_runner()` calls in 823 tests, with each taking ~0.3s. The reload proportion is probably artificially high given the choice of tests, but still probably worth addressing.

> Why we need to redo those things when protocol change?

I don't know. All of the URL changes I saw were between:

```
http://web-platform.test:8001/testharness_runner.html
https://web-platform.test:8444/testharness_runner.html
```

Reply:
That might not be needed these days. It used to be required that the test runner was same-origin with the test page in order to allow communicating between them using `postMessage`. But I _think_ we don't actually require it to be same origin any more (I didn't check), so it might be sufficient to do nothing.